File: 0648-replace-words/0648-replace-words.js
/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
var replaceWords = function(dictionary, sentence) {
    dictionary.sort();
    let arr = sentence.split(" ");
    for(let i = 0; i < arr.length; i++) {
        let ele = arr[i];
        let charArr = ele.split("");
        let builtStr = "";
        for(let char of charArr) {
            builtStr += char;
            if(dictionary.includes(builtStr)){
                arr[i] = builtStr;
                break;
            }
        }
    }
    let ans = "";
    for(let ele of arr) {
        ans += ele + " ";
    }
    return ans.trim();
};

File: 0648-replace-words/0648-replace-words.java
class Solution {
    public String replaceWords(List<String> dictionary, String sentence) {
        TreeSet<String> set = new TreeSet<>();
        for(String str : dictionary) {
            set.add(str);
        }
        String arr[] = sentence.split(" ");
        int n = arr.length;
        for(int i = 0; i < n; i++) {
            String s = arr[i];
            String curr = "";
            for(char c: s.toCharArray()) {
                curr += c;
                // System.out.println(curr);
                if(set.contains(curr)) {
                    arr[i] = curr;
                    break;
                }
                
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for(String s : arr) {
            sb.append(s);
            sb.append(" ");
        }
        return sb.toString().trim();
    }
}